logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultPathAssertions.kt

```diff
@@ -194,4 +193,31 @@ class DefaultPathAssertions : PathAssertions {
                     container.resolve(entry).collect { _logicAppend { exists(NOFOLLOW_LINKS) } }
                 }
         ).build()
+
+    override fun <T : Path> isEmptyDirectory(container: AssertionContainer<T>): Assertion =
+        if (container.isDirectory().holds()) {
+            container.changeSubject.unreported {
+                it.runCatchingIo { Files.list(it).findAny() }
+            }
+                .let { expectResult ->
+                    assertionBuilder.descriptive.withTest(expectResult)
+                    { it is Success && it.value.isEmpty }
+                        .withFailureHintBasedOnDefinedSubject(expectResult) {
+                            when (it) {
+                                is Success ->
+                                    assertionBuilder.descriptive.failing
+                                        .withDescriptionAndRepresentation(
+                                            DIRECTORY_HAS_FILE,
+                                            it.value.orElseThrow().toString()
+                                        )
+                                        .build()
+                                is Failure -> hintForIoException(it.path, it.exception)
+                            }
+                        }
+                        .withDescriptionAndRepresentation(DescriptionBasic.IS, AN_EMPTY_DIRECTORY)
+                        .build()
+                }
+        } else {
+            container.isDirectory()
+        }
```

Comment:
If a file is not a directory, the current logic will check whether the file is a directory twice: First in the if-condition, and then when displaying the assertion created by `container.isDirectory()` to the user. This is wasteful in general, but even worse for file system operations: The file could be changed between checking the condition and testing the second assertion. In that case, users would get confusing messages.

Luckily, I think there is an easier solution here: First, you do not need to test whether the file is a directory. Instead, you can simply add the assertion for the file being a directory. If it’s not, the whole assertion will already fail. You then add the second assertion to the `AssertionContainer`, like you already do in the `let`-block. Only now you will need to be aware that the second assertion’s test may also be called if the first assertion fails. This is not too bad, though, you simply need to handle the corresponding exception.

Reply:
It seems like the file could always be changed even if we have 2 assertions? What happens if the first assertion holds but by the time the second assertion's test is called, the path ceased to be a directory and we get an exception, we could either
- ignore this exception which means the user gets no error even though the path probably does not point to an empty directory
- return isDirectory() again, thought this presents the same issue as before so probably not
- build our own failing assertion and return it to the user, though I don't know how to format it in the same way as isDirectory()? If we do this I think we might as well not use isDirectory() and so that the error formatting is consistent.

Let me know if there's something I'm missing and/or which approach should I take.

Reply:
None of are file assertions are atomic, that is true. However, the impact of parallel modifications can be more or less bad. I’d argue that the current implementation can lead to irritating results, while the implementations I have in mind and @robstoll suggested should not be a problem, because the user can understand the reporting and deduce what happened.